File: front-end/src/types/stock.ts

import type { AppDocument } from "./document";

export type StockUnit =
  | "UNITE"
  | "BOITE"
  | "PAQUET"
  | "LITRE"
  | "KILOGRAMME"
  | "METRE";

export type SupplyRequestStatus =
  | "PENDING"
  | "APPROVED"
  | "REJECTED"
  | "RECEIVED";

export type StockMovementType =
  | "ENTRY"
  | "EXIT";

export type RestockAlertStatus =
  | "WAITING"
  | "READY_NOTIFIED";

export interface StockArticle {
  id: number;

  reference: string;
  barcode: string;
  brand: string;

  designation: string;
  designationAr: string;

  category: string;
  categoryAr: string;

  quantity: number;
  minimumQuantity: number;

  unit: StockUnit;

  location: string;
  locationAr: string;

  unitPriceHt: number;
  vatRate: number;

  documents: AppDocument[];

  updatedAt: string;
}

export interface StockArticleFormData {
  reference: string;
  barcode: string;
  brand: string;

  designation: string;
  designationAr: string;

  category: string;
  categoryAr: string;

  quantity: number;
  minimumQuantity: number;

  unit: StockUnit;

  location: string;
  locationAr: string;

  unitPriceHt: number;
  vatRate: number;

  documents: AppDocument[];
}

export interface StockMovement {
  id: number;

  articleId: number;

  articleDesignation: string;
  articleDesignationAr: string;

  type: StockMovementType;

  quantity: number;
  reason: string;

  supplierOrBeneficiary?: string;
  reference?: string;

  performedBy: string;
  date: string;

  unitPriceHt: number;
  vatRate: number;

  documents: AppDocument[];

  supplyRequestId?: number;
}

export interface SupplyRequest {
  id: number;

  articleId: number;
  articleDesignation: string;
  articleDesignationAr: string;

  requestedQuantity: number;

  requesterId: number;
  requester: string;

  reason: string;
  requestDate: string;

  status: SupplyRequestStatus;

  rejectionReason?: string;
  decisionDate?: string;
  receivedAt?: string;

  documents: AppDocument[];
}

export interface RestockAlert {
  id: number;

  articleId: number;
  articleDesignation: string;
  articleDesignationAr: string;

  requestedQuantity: number;
  availableQuantityAtRequest: number;

  requesterId: number;
  requester: string;

  reason: string;
  createdAt: string;

  status: RestockAlertStatus;
  readyNotifiedAt?: string;
}

export interface StockFilters {
  search: string;
  category: string;
  alertOnly: boolean;
}

export function calculateUnitPriceTtc(
  unitPriceHt: number,
  vatRate: number,
) {
  return unitPriceHt * (1 + vatRate / 100);
}

export function calculateTotalHt(
  quantity: number,
  unitPriceHt: number,
) {
  return quantity * unitPriceHt;
}

export function calculateTotalTtc(
  quantity: number,
  unitPriceHt: number,
  vatRate: number,
) {
  return quantity * calculateUnitPriceTtc(
    unitPriceHt,
    vatRate,
  );
}

File: mobile/lib/models/stock.dart

import 'app_document.dart';

enum StockUnit {
  unite,
  boite,
  paquet,
  litre,
  kilogramme,
  metre,
}

enum MovementType {
  entry,
  exit,
}

enum SupplyRequestStatus {
  pending,
  approved,
  rejected,
  received,
}

enum RestockAlertStatus {
  waiting,
  readyNotified,
}

class StockArticle {
  StockArticle({
    required this.id,
    required this.reference,
    required this.barcode,
    required this.brand,
    required this.designation,
    required this.designationAr,
    required this.category,
    required this.categoryAr,
    required this.quantity,
    required this.minimumQuantity,
    required this.unit,
    required this.location,
    required this.locationAr,
    required this.unitPriceHt,
    required this.vatRate,
    required this.updatedAt,
    required this.documents,
  });

  final int id;

  String reference;
  String barcode;
  String brand;

  String designation;
  String designationAr;

  String category;
  String categoryAr;

  int quantity;
  int minimumQuantity;

  StockUnit unit;

  String location;
  String locationAr;

  double unitPriceHt;
  double vatRate;

  DateTime updatedAt;

  List<AppDocument> documents;

  bool get isLowStock =>
      quantity <= minimumQuantity;

  double get unitPriceTtc =>
      unitPriceHt *
      (1 + vatRate / 100);

  double get totalHt =>
      quantity * unitPriceHt;

  double get totalTtc =>
      quantity * unitPriceTtc;
}

class StockMovement {
  StockMovement({
    required this.id,
    required this.articleId,
    required this.articleDesignation,
    required this.articleDesignationAr,
    required this.type,
    required this.quantity,
    required this.reason,
    required this.performedBy,
    required this.date,
    required this.unitPriceHt,
    required this.vatRate,
    required this.documents,
    this.supplierOrBeneficiary,
    this.reference,
    this.supplyRequestId,
  });

  final int id;
  final int articleId;

  final String articleDesignation;
  final String articleDesignationAr;

  final MovementType type;
  final int quantity;

  final String reason;
  final String performedBy;

  final DateTime date;

  final double unitPriceHt;
  final double vatRate;

  final List<AppDocument> documents;

  final String? supplierOrBeneficiary;
  final String? reference;
  final int? supplyRequestId;

  double get unitPriceTtc =>
      unitPriceHt *
      (1 + vatRate / 100);

  double get totalHt =>
      quantity * unitPriceHt;

  double get totalTtc =>
      quantity * unitPriceTtc;
}

class SupplyRequest {
  SupplyRequest({
    required this.id,
    required this.articleId,
    required this.articleDesignation,
    required this.articleDesignationAr,
    required this.requestedQuantity,
    required this.requesterId,
    required this.requester,
    required this.reason,
    required this.requestDate,
    required this.status,
    required this.documents,
    this.rejectionReason,
    this.decisionDate,
    this.receivedAt,
  });

  final int id;

  final int articleId;
  final String articleDesignation;
  final String articleDesignationAr;

  final int requestedQuantity;

  final int requesterId;
  final String requester;

  final String reason;
  final DateTime requestDate;

  SupplyRequestStatus status;

  String? rejectionReason;
  DateTime? decisionDate;
  DateTime? receivedAt;

  final List<AppDocument> documents;
}

class RestockAlert {
  RestockAlert({
    required this.id,
    required this.articleId,
    required this.articleDesignation,
    required this.articleDesignationAr,
    required this.requestedQuantity,
    required this.availableQuantityAtRequest,
    required this.requesterId,
    required this.requester,
    required this.reason,
    required this.createdAt,
    required this.status,
    this.readyNotifiedAt,
  });

  final int id;

  final int articleId;
  final String articleDesignation;
  final String articleDesignationAr;

  final int requestedQuantity;
  final int availableQuantityAtRequest;

  final int requesterId;
  final String requester;

  final String reason;
  final DateTime createdAt;

  RestockAlertStatus status;
  DateTime? readyNotifiedAt;
}
